bluefin-internal/src/Bluefin/Internal/Examples.hs

```diff
@@ -573,17 +581,52 @@ runFileSystemIO ex io k =
         Left e -> throw ex (show e)
         Right r -> pure r
 
+runFileSystemPure' ::
+  (e1 :> es) =>
+  Exception String e1 ->
+  [(FilePath, String)] ->
+  (forall e2. Handle FileSystem e2 -> Eff (e2 :& es) r) ->
+  Eff es r
+runFileSystemPure' ex fs0 k =
+  evalState fs0 $ \fs ->
+    -- This extra handler confirms that we can create our handle in
+    -- the context of more than one handler
+    evalState () $ \_ ->
+      within
+        k
+        MkFileSystem
+          { readFileImpl = \path -> do
+              fs' <- get fs
+              case lookup path fs' of
+                Nothing ->
+                  throw ex ("File not found: " <> path)
+                Just s -> pure s,
+            writeFileImpl = \path contents ->
+              modify fs ((path, contents) :)
+          }
+
 action :: (e :> es) => FileSystem e -> Eff es String
 action fs = do
   file <- readFile fs "/dev/null"
   when (length file == 0) $ do
     writeFile fs "/tmp/bluefin" "Hello!\n"
   readFile fs "/tmp/doesn't exist"
 
+action' :: e :> es => Handle FileSystem e -> Eff es String
+action' fs = do
+  file <- readFileImpl fs "/dev/null"
+  when (length file == 0) $ do
+    writeFileImpl fs "/tmp/bluefin" "Hello!\n"
+  readFileImpl fs "/tmp/doesn't exist"
+
 exampleRunFileSystemPure :: Either String String
 exampleRunFileSystemPure = runPureEff $ try $ \ex ->
   runFileSystemPure ex [("/dev/null", "")] action
 
+-- exampleRunFileSystemPure' :: Either String String
+-- exampleRunFileSystemPure' = runPureEff $ try $ \ex ->
+--   runFileSystemPure' ex [("/dev/null", "")] action'
```

Comment:
Ah, but this is no good :( I get an ambiguous type error if I try to do this. So maybe we need to make `Handle` a newtype? In which case we lose the nice property that we can use the record fields without further adjustment.

Reply:
Good catch. I should have guessed that `Handle` works badly with type inference. Personally if that doesn't work I would rather go back to monomorphic records (i.e., replace `Handle f e` with `f e` and ask users to wrap method invocations with `useImpl`, this is commonplace in other effect systems (`perform` in OCaml, `send` in effectful)) (but then we should reevaluate how this PR compares with the similar feature that's already in bluefin, because that `Handle` trick was one of this PR's main selling points).

I'm not a fan of the following hacks, but they could be attempted if one really wanted to make `Handle` work with type inference. The common idea is that we need the effect `e` to appear not only in constraints in `Handle`, but somehow as an index of a data type.

- Hack 1: users should declare records with an extra phantom effect parameter:

    ```haskell
    data FileSystem e es = FileSystem {
      readFile :: String -> Eff es String,
      writeFile :: String -> String -> Eff es ()
    }
    type Handle f e = forall es. e :> es => f e es
    ```

- Hack 2: if we want users to not think about this hack, `Handle`  should somehow introduce that phantom parameter. And in order to not lose the field accessors (because that's the whole point of `Handle`), we do some `HasField` shenanigans.

    ```haskell
    newtype TaggedHandle f e es = TaggedHandle (f es)
    instance HasField name (f es) a => HasField name (TaggedHandle f e es) a where
      getField (TaggedHandle r) = getField @name r

    type Handle f e = forall es. e :> es => TaggedHandle f e es
    ```